Automatic recognition of hex colour strings in layer data

zindy opened the pull request:
# Description
<!-- What does this pull request (PR) do? Why is it necessary? -->
<!-- Tell us about your new feature, improvement, or fix! -->
<!-- If your change includes user interface changes, please add an image, or an animation "An image is worth a thousand words!" -->
<!-- You can use https://www.cockos.com/licecap/ or similar to create animations -->

This PR adds the automatic recognition of hex colour strings in layer data and if necessary, generates a linear colormap from black to that colour and adds it to the list of available colormaps. If the hex colour string defines a bright {white | red | green | blue | cyan | magenta | yellow} colour, then the corresponding colormap will be used instead of adding a new one.

My goal with this fix is to simplify colour handling in reader plugins (I especially had [napari-czifile2](https://github.com/BodenmillerGroup/napari-czifile2) in mind) by being able to just specify a channel colour as a hex string and let napari generate (or reuse) a colormap for this particular colour.

# References
<!-- What resources, documentation, and guides were used in the creation of this PR? -->

My reference was the [colormap_utils.py](https://github.com/napari/napari/blob/main/napari/utils/colormaps/colormap_utils.py) source code and the modules it imports. I noticed that `transform_color()` in [napari.utils.colormaps.standardize_color](https://github.com/napari/napari/blob/main/napari/utils/colormaps/standardize_color.py) already had the ability to convert hex colour strings.

I first mentioned this work on the [image.sc forum](https://forum.image.sc/t/napari-development-question-about-colormaps-in-the-image-layer/84264) and got one favourable return (thanks Juan!).

## Type of change
<!-- Please delete options that are not relevant. -->
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

TODO: Possibly the last point is to mention the new capability in the documentation.

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
I used [napari-czifile2](https://github.com/BodenmillerGroup/napari-czifile2) to check my addition with the following code (in Jupyter):
```python
import napari_czifile2
import napari
%gui qt

# Read the layer data (one for each scene in the CZI file) and extract the first scene.
fn = 'somefile.czi'
layer = napari_czifile2._reader.reader_function(fn)[0]

# Check various combinations of colours:
layer[1]['colormap']=['#f0F','#fdff00fF']

# Open a viewer and add the layer as image + parameters:
viewer=napari.Viewer()
viewer.add_image(layer[0], **layer[1])
```

### Things I checked:
- #rgb #rgba #rrggbb #rrggbbaa combinations
- used in combination with `None` in the colormap list
- case (in)sensitivity
- one of the colours for which there is a colormap already (r,g,b,c,m,y,w)
- custom colours and how their name look in the napari viewer colormap list
- malformed hex colours
- missing or extra elements in colormap (extra is fine, missing is not)

### In particular:

#### Missing colormap:
```
IndexError: Error adding multichannel image with data shape (1, 121, 2, 1024, 1024).
Requested channel_axis (2) had length 2, but the 'colormap' argument only provided 1 values. 
```

#### Malformed colormap
The existing exception error message should be enough to understand what happened (hex colour string used was `'#ff'`) :
```
KeyError: 'Colormap "#ff" not found in either vispy or matplotlib. Recognized colormaps are: "GrBu", "GrBu_d", "PiYG", "PuGr", "RdBu", "RdYeBuCy", "autumn", "blues", "cool", "coolwarm", "cubehelix", "diverging", "fire", "gist_earth", "gray", "gray_r", "grays", "greens", "hot", "hsl", "hsv", "husl", "ice", "inferno", "light_blues", "magma", "orange", "plasma", "reds", "single_hue", "spring", "summer", "turbo", "twilight", "twilight_shifted", "viridis", "winter"'
```

#### What generated custom colormap names looks like in the list:
The new names start with `custom-` and display a lowercase version of the hex color string as it was passed to the viewer:
![image](https://github.com/napari/napari/assets/3523902/f5446177-cd90-42ac-8131-3d94b7023552)

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

I used existing napari code wherever possible, so hopefully the PR is the minimum amount of code needed for the desired functionality.

I have tested the code pretty extensively but need to check if there are unit tests for colormap-utils and if there are, add some of my own.

I haven't checked yet where the documentation needs amending but certainly can if you would like to add this capability to napari.

Cheers,
Egor

zindy committed:
hexcolor strings conversion into colormaps

Added the automatic recognition of a hex color strings and conversion to a linear colormap from black to that color

zindy committed:
Merge branch 'napari:main' into hex-colors

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

zindy committed:
Fix a typing error

zindy committed:
Fix indentation

zindy committed:
Fix indentation

zindy committed:
Merge branch 'main' into hex-colors

Czaki commented on napari/utils/colormaps/colormap_utils.py:
```diff
@@ -745,14 +765,64 @@ def ensure_colormap(colormap: ValidColormapArg) -> Colormap:
+    if name is None:
+        name = 'custom-' + hexcolor[1:].lower()
+
+    return name
```
```suggestion
    return rgba_colormaps.get(hexcolor_rgba.lower(), 'custom-' + hexcolor[1:].lower())
```

This part could be simplified

zindy replied:
Thank you, this is much nicer!

Czaki commented on napari/utils/colormaps/colormap_utils.py:
```diff
@@ -618,7 +618,27 @@ def ensure_colormap(colormap: ValidColormapArg) -> Colormap:
+                            )
+                        )
+            else:
+                name = colormap
```
I think that you should use here `vispy.color.Color` to recognize any string that is a valid vispy color. Maybe something like this (plus handle vispy_or_mpl_colormap).

```suggestion
            cmap = AVAILABLE_COLORMAPS.get(colormap, AVAILABLE_COLORMAPS.get(f"custom-{colormap}", None)
            if cmap is None:
                cmap = _colormap_from_colors(colormap, name=f"custom-{colormap}")
                for cmap_ in AVAILABLE_COLORMAPS.values():
                    if cmap_.controls == cmap.controls and cmap_.colors == cmap.colors and cmap_.interpolation == cmap.interpolation:
                        cmap = cmap_
                        break
            name = cmap.name
            AVAILABLE_COLORMAPS[name] = cmap
```

zindy replied:
@Czaki thank you very much for the idea, indeed this makes my _hexcolor_to_name() function completely unnecessary as you are directly testing the colors of the newly generated cmap against any of the existing ones.

There may be a slight issue with the way the `.colors` attribute is defined as a ColorArray of a single nx4 numpy array rather than a list (?) of colors (like [in the vispy example given for ColorArray](https://www.ssec.wisc.edu/~davidh/tmp/vispy/api/vispy.color.color_array.html#vispy.color.color_array.ColorArray), which messes up your `cmap_.colors == cmap.colors` test, in the sense that an array of bools, one for each component is returned rather than a single true/false value.

So instead this worked for me (testing shapes are equal and using `np.all()`):
```python
                for cmap_ in AVAILABLE_COLORMAPS.values():
                    if np.array_equal(cmap_.controls, cmap.controls) and cmap_.colors.shape == cmap.colors.shape and np.all(cmap_.colors == cmap.colors) and cmap_.interpolation == cmap.interpolation:
                        cmap = cmap_
                        break
```

And here's how I've tested it so far:
```python
import napari.utils.colormaps.colormap_utils as cu
cmap = cu.ensure_colormap('#f0fF')
cu.AVAILABLE_COLORMAPS.get(cmap.name).name
```
returns `'magenta'`

I will check this some more and amend the code as necessary.

Cheers,
Egor

Czaki reviewed (commented):
I think that code could be simplified and much more generic to handle and recognizable color by vispy

zindy committed:
Code simplification

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

zindy committed:
Code clean-up as per @Czaki suggestion

Suggestion was made here by @Czaki https://github.com/napari/napari/pull/6102#discussion_r1278287232 to use vispy.color.Color to recognize any string that is a valid vispy color.

This now allows strings like 'orange' which are valid colour names.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki commented on napari/utils/colormaps/colormap_utils.py:
```diff
@@ -618,27 +618,26 @@ def ensure_colormap(colormap: ValidColormapArg) -> Colormap:
+                        trans._(
+                            'Unrecognized Colormap String Value: {name}',
+                            deferred=True,
+                            name=colormap,
```
here should be the call of `vispy_or_mpl_colormap` instead of raising an exception.

zindy replied:
Thanks, this makes a lot of sense and the generated error message is perfectly understandable:
```python
import napari.utils.colormaps.colormap_utils as cu
cmap = cu.ensure_colormap('some_name')
```
returns
```
KeyError: 'Colormap "some_name" not found in either vispy or matplotlib. Recognized colormaps are: "GrBu", "GrBu_d", "PiYG", "PuGr", "RdBu", "RdYeBuCy", "autumn", "blues", "cool", "coolwarm", "cubehelix", "diverging", "fire", "gist_earth", "gray", "gray_r", "grays", "greens", "hot", "hsl", "hsv", "husl", "ice", "inferno", "light_blues", "magma", "orange", "plasma", "reds", "single_hue", "spring", "summer", "turbo", "twilight", "twilight_shifted", "viridis", "winter"'
```
Any of the suggested names will work, for example 'twilight_shifted' will generate quite a complex colormap.

Let me know if anything else needs amending, and thank you for your reviews!

Cheers,
Egor

zindy committed:
replaced exception with call to vispy_or_mpl_colormap

brisvag reviewed (commented):
Looks great! Small suggestion: maybe it would be nicer for the custom colormap name to contain the `#` at the beginning (that's more immediately recognizable as hex). Other than that, some simple tests as mentioned by @Czaki and we're good to go :)

zindy committed:
Merge branch 'main' into hex-colors

zindy committed:
added tests for ensure_colormap with hex color strings

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki commented on napari/utils/colormaps/colormap_utils.py:
```diff
@@ -618,7 +618,30 @@ def ensure_colormap(colormap: ValidColormapArg) -> Colormap:
+                        cmap = cmap_
+                        break
+
+            name = cmap.name
```
here you need to not use `cmap` but some other name (that will be assigned to `cmap` at the end. The line 622 gives type `Optional[Colormap]` and mypy is not able to recognize that it is no longer true after line 631.

Czaki commented on napari/utils/colormaps/colormap_utils.py:
```diff
@@ -618,7 +618,30 @@ def ensure_colormap(colormap: ValidColormapArg) -> Colormap:
+
+                for cmap_ in AVAILABLE_COLORMAPS.values():
+                    if (
+                        np.array_equal(cmap_.controls, cmap.controls)
```
This line requires fixing type annotation of `napari.utils.colormaps.colormap.Colormap` (field `controls`). As the validators of this field, ensure that this field is always Array, but type annotation state that it is `Optional[Array]`

zindy replied:
This one I have trouble understanding, sorry.

Let me know if this is acceptable: I've modified my test starting at line 633 to check whether either of `cmap_.controls` or `cmap.controls` (actually renamed `custom_cmap` as per explanation above) is `None` before my `np.equal` condition.

```python
                for cmap_ in AVAILABLE_COLORMAPS.values():
                    if (
                        cmap_.controls is not None
                        and custom_cmap.controls is not None
                        and np.array_equal(cmap_.controls, custom_cmap.controls)
                        and cmap_.colors.shape == custom_cmap.colors.shape
                        and np.all(cmap_.colors == custom_cmap.colors)
                        and cmap_.interpolation == custom_cmap.interpolation
                    ):
                        custom_cmap = cmap_
                        break
```

Czaki replied:
change line 54 of `napari/utils/colormaps/colormap.py` to:

```python
controls: Array = Field(default_factory=lambda :Array([]))
```

zindy committed:
attempt to fix incompatible type erros

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

zindy committed:
make Colormap.controls default an empty array instead of None

This change was suggested by @Czaki here: https://github.com/napari/napari/pull/6102#discussion_r1280666063

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

zindy committed:
simplified code for colormap controls comparison

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki commented on napari/utils/colormaps/colormap.py:
```diff
@@ -3,7 +3,7 @@
 from typing import Optional
```
```suggestion
from typing import cast, Optional
```

Czaki commented on napari/utils/colormaps/colormap.py:
```diff
@@ -51,7 +51,7 @@ class Colormap(EventedModel):
     _display_name: Optional[str] = PrivateAttr(None)
     interpolation: ColormapInterpolationMode = ColormapInterpolationMode.LINEAR
-    controls: Optional[Array] = None
+    controls: Array = Field(default_factory=lambda: Array([]))
```
```suggestion
    controls: Array = Field(default_factory=lambda: cast(Array, []))
```

Czaki reviewed (commented):
I use wrong version of numpy. This should work

zindy committed:
Update napari/utils/colormaps/colormap.py

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

zindy committed:
Update napari/utils/colormaps/colormap.py

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

zindy committed:
Update colormap.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

zindy committed:
Merge branch 'main' into hex-colors

Czaki commented on napari/utils/colormaps/colormap_utils.py:
```diff
@@ -618,7 +618,30 @@ def ensure_colormap(colormap: ValidColormapArg) -> Colormap:
+            AVAILABLE_COLORMAPS[name] = custom_cmap
             if name not in AVAILABLE_COLORMAPS:
                 cmap = vispy_or_mpl_colormap(
                     name
```
this lines looks obsolete, as function is called in line 631

zindy replied:
Very well spotted, thanks.

zindy committed:
Unused code, added custom name clean-up

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

brisvag commented on napari/utils/colormaps/_tests/test_colormaps.py:
```diff
@@ -235,3 +235,33 @@ def test_ensure_colormap_with_multi_colors(colors):
+        assert cmap.name == 'custom-' + color.lower()
+
+
+def test_ensure_colormap_with_recognized_hex_color_string():
```
Would be nicer to have these tests use `parametrize` instead of a loop, but that's not a blocker for merging :)

zindy replied:
> Would be nicer to have these tests use `parametrize` instead of a loop,

No worries, that was an easy change to make :)

And I added one extra test to make sure that when a matplotlib colour name (like 'chartreuse') is passed to ensure_colormap(), the name of the generated colormap is identical to that.

brisvag commented on napari/utils/colormaps/colormap_utils.py:
```diff
@@ -618,12 +618,34 @@ def ensure_colormap(colormap: ValidColormapArg) -> Colormap:
+                    f"custom-{colormap.lower()}"
+                    if '#' in colormap
+                    else colormap
+                )
```
A bit more robust:

```suggestion
                name = (
                    f"custom-{colormap.lower()}"
                    if colormap.startswith('#')
                    else colormap
                )
```

zindy committed:
More robust hex color string recognition

Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>

zindy committed:
parametrize and added one more test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

zindy committed:
Merge branch 'main' into hex-colors

zindy committed:
removed the 'custom-' prefix of the new colormap name

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

zindy committed:
Update test_colormaps.py to reflect change in colormap name

zindy committed:
Merge branch 'main' into hex-colors
